File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.java
package org.ccci.gto.android.common.db;

import android.os.Parcel;
import android.os.Parcelable;
import android.util.Pair;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.ccci.gto.android.common.util.ArrayUtils;

import java.util.Arrays;

public abstract class Expression extends ShimExpression {
    static final String[] NO_ARGS = new String[0];

    Expression() {
    }

    protected int numOfArgs() {
        return getNumOfArgs();
    }

    @NonNull
    public  Expression args(@NonNull final String... args) {
        super.args(args);
        return this;
    }

    @NonNull
    public final Binary and(@NonNull final Expression expression) {
        return binaryExpr(Binary.AND, expression);
    }

    @NonNull
    public final Binary eq(@NonNull final Number constant) {
        return eq(constant(constant));
    }

    @NonNull
    public final Binary eq(@NonNull final String constant) {
        return eq(constant(constant));
    }

    @NonNull
    public final Binary eq(@NonNull final Object constant) {
        return eq(constant(constant));
    }

    @NonNull
    public final Binary eq(@NonNull final Expression expression) {
        return binaryExpr(Binary.EQ, expression);
    }

    @NonNull
    public final Binary lt(@NonNull final Number constant) {
        return lt(constant(constant));
    }

    @NonNull
    public final Binary lt(@NonNull final Object constant) {
        return lt(constant(constant));
    }

    @NonNull
    public final Binary lt(@NonNull final Expression expression) {
        return binaryExpr(Binary.LT, expression);
    }

    @NonNull
    public final Binary lte(@NonNull final Number constant) {
        return lte(constant(constant));
    }

    @NonNull
    public final Binary lte(@NonNull final Object constant) {
        return lte(constant(constant));
    }

    @NonNull
    public final Binary lte(@NonNull final Expression expression) {
        return binaryExpr(Binary.LTE, expression);
    }

    @NonNull
    public final Binary gt(@NonNull final Number constant) {
        return gt(constant(constant));
    }

    @NonNull
    public final Binary gt(@NonNull final Object constant) {
        return gt(constant(constant));
    }

    @NonNull
    public final Binary gt(@NonNull final Expression expression) {
        return binaryExpr(Binary.GT, expression);
    }

    @NonNull
    public final Binary gte(@NonNull final Number constant) {
        return gte(constant(constant));
    }

    @NonNull
    public final Binary gte(@NonNull final Object constant) {
        return gte(constant(constant));
    }

    @NonNull
    public final Binary gte(@NonNull final Expression expression) {
        return binaryExpr(Binary.GTE, expression);
    }

    @NonNull
    public final Binary in(@NonNull final Expression... expressions) {
        return new Binary(Binary.IN, ArrayUtils.merge(Expression.class, new Expression[] {this}, expressions));
    }

    @NonNull
    public final Binary notIn(@NonNull final Expression... expressions) {
        return new Binary(Binary.NOTIN, ArrayUtils.merge(Expression.class, new Expression[] {this}, expressions));
    }

    @NonNull
    public final Binary is(@NonNull final Expression expression) {
        return binaryExpr(Binary.IS, expression);
    }

    @NonNull
    public final Binary isNot(@NonNull final Expression expression) {
        return binaryExpr(Binary.ISNOT, expression);
    }

    @NonNull
    public final Binary or(@NonNull final Expression expression) {
        return binaryExpr(Binary.OR, expression);
    }

    @NonNull
    public final Binary ne(@NonNull final Number constant) {
        return ne(constant(constant));
    }

    @NonNull
    public final Binary ne(@NonNull final String constant) {
        return ne(constant(constant));
    }

    @NonNull
    public final Binary ne(@NonNull final Object constant) {
        return ne(constant(constant));
    }

    @NonNull
    public final Binary ne(@NonNull final Expression expression) {
        return binaryExpr(Binary.NE, expression);
    }

    @NonNull
    protected Binary binaryExpr(@NonNull final String op, @NonNull final Expression expression) {
        return new Binary(op, this, expression);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(final Parcel out, final int flags) {
    }

    public static class Binary extends Expression {
        static final String LT = "<";
        static final String LTE = "<=";
        static final String GT = ">";
        static final String GTE = ">=";
        static final String EQ = "==";
        static final String NE = "!=";
        static final String IS = "IS";
        static final String ISNOT = "IS NOT";
        static final String IN = "IN";
        static final String NOTIN = "NOT IN";
        static final String AND = "AND";
        static final String OR = "OR";

        @NonNull
        private final String mOp;
        @NonNull
        private final Expression[] mExprs;
        private final int mNumOfArgs;

        @Nullable
        private transient Pair<String, String[]> mSql;

        Binary(@NonNull final String op, @NonNull final Expression... exprs) {
            mOp = op;
            mExprs = exprs;
            mNumOfArgs = calcNumOfArgs();
        }

        Binary(@NonNull final Parcel in, @Nullable final ClassLoader loader) {
            mOp = in.readString();
            final Parcelable[] exprs = in.readParcelableArray(loader);
            mExprs = new Expression[exprs.length];
            for (int i = 0; i < exprs.length; i++) {
                mExprs[i] = (Expression) exprs[i];
            }
            mNumOfArgs = calcNumOfArgs();
        }

        private int calcNumOfArgs() {
            int sum = 0;
            for (final Expression expr : mExprs) {
                sum += expr.numOfArgs();
            }
            return sum;
        }

        @NonNull
        @Override
        @SuppressWarnings("checkstyle:MissingSwitchDefault")
        protected Binary binaryExpr(@NonNull final String op, @NonNull final Expression expression) {
            // chain binary expressions together when possible
            switch (mOp) {
                case AND:
                case OR:
                    if (mOp.equals(op)) {
                        return new Binary(mOp,
                                          ArrayUtils.merge(Expression.class, mExprs, new Expression[] {expression}));
                    }
            }

            return super.binaryExpr(op, expression);
        }

        @NonNull
        @Override
        public KotlinExpression not() {
            // sometimes we can just change our own op for not()
            switch (mOp) {
                case EQ:
                    return new Binary(NE, mExprs);
                case NE:
                    return new Binary(EQ, mExprs);
                case IS:
                    return new Binary(ISNOT, mExprs);
                case ISNOT:
                    return new Binary(IS, mExprs);
                case IN:
                    return new Binary(NOTIN, mExprs);
                case NOTIN:
                    return new Binary(IN, mExprs);
                default:
                    return super.not();
            }
        }

        @Override
        protected int numOfArgs() {
            return mNumOfArgs;
        }

        @NonNull
        @Override
        public Binary args(@NonNull final String... args) {
            if (args.length != mNumOfArgs) {
                throw new IllegalArgumentException("incorrect number of args specified");
            }

            // short-circuit if there are no args
            if (args.length == 0) {
                return this;
            }

            int pos = 0;
            final Expression[] exprs = new Expression[mExprs.length];
            for (int i = 0; i < mExprs.length; i++) {
                final int num = mExprs[i].numOfArgs();
                exprs[i] = num > 0 ? (Expression) mExprs[i].args(Arrays.copyOfRange(args, pos, pos + num)) : mExprs[i];
                pos += num;
            }
            return new Binary(mOp, exprs);
        }

        @NonNull
        @Override
        public Pair<String, String[]> buildSql(@NonNull final AbstractDao dao) {
            // generate SQL if it hasn't been generated yet
            if (mSql == null) {
                int i = 0;
                final StringBuilder sql = new StringBuilder();
                String[] args = NO_ARGS;
                sql.append('(');

                final boolean isIn = IN.equals(mOp) || NOTIN.equals(mOp);
                if (isIn) {
                    // "{mExpr[0]} IN ("
                    final Pair<String, String[]> resp = mExprs[0].buildSql(dao);
                    sql.append(resp.first);
                    args = ArrayUtils.merge(String.class, args, resp.second);
                    sql.append(' ').append(mOp).append(" (");
                    i++;
                }

                // "{mExpr[i]} {mOp} {mExpr[i+1]} ..."
                boolean first = true;
                for (; i < mExprs.length; i++) {
                    final Expression expr = mExprs[i];
                    if (!first) {
                        sql.append(' ').append(isIn ? ',' : mOp).append(' ');
                    }
                    final Pair<String, String[]> resp = expr.buildSql(dao);
                    sql.append(resp.first);
                    args = ArrayUtils.merge(String.class, args, resp.second);
                    first = false;
                }

                if (isIn) {
                    sql.append(')');
                }

                sql.append(')');
                mSql = Pair.create(sql.toString(), args);
            }

            return mSql;
        }

        @Override
        public void writeToParcel(final Parcel out, final int flags) {
            super.writeToParcel(out, flags);
            out.writeString(mOp);
            out.writeParcelableArray(mExprs, 0);
        }

        public static final Creator<Binary> CREATOR = new BinaryExpressionCreator();

        private static class BinaryExpressionCreator implements ClassLoaderCreator<Binary> {
            @Override
            public Binary createFromParcel(@NonNull final Parcel in) {
                return new Binary(in, null);
            }

            @Override
            public Binary[] newArray(final int size) {
                return new Binary[size];
            }

            @Override
            public Binary createFromParcel(@NonNull final Parcel in, @Nullable final ClassLoader loader) {
                return new Binary(in, loader);
            }
        }
    }
}

File: gto-support-db/src/main/java/org/ccci/gto/android/common/db/Expression.kt
package org.ccci.gto.android.common.db

import android.os.Parcelable
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import org.ccci.gto.android.common.db.AbstractDao.Companion.bindValues

sealed class Expression : Parcelable {
    companion object {
        @JvmField
        val NULL = Literal(isConstant = true)

        @JvmStatic
        fun bind() = Literal(isConstant = false)
        @JvmStatic
        fun bind(value: Any) = Literal(bindValues(value)[0], false)
        @JvmStatic
        fun bind(value: Number) = Literal(value, false)
        @JvmStatic
        fun bind(value: String) = Literal(value, false)

        @JvmStatic
        fun constant(value: Any) = Literal(bindValues(value)[0], true)
        @JvmStatic
        fun constant(value: Number) = Literal(value, true)
        @JvmStatic
        fun constant(value: String) = Literal(value, true)

        @JvmStatic
        fun constants(vararg values: Any) = constants(*bindValues(*values))
        @JvmStatic
        fun constants(vararg values: Number) = Array(values.size) { constant(values[it]) }
        @JvmStatic
        fun constants(vararg values: String) = Array(values.size) { constant(values[it]) }

        @JvmStatic
        fun field(name: String) = Field(name = name)

        @JvmStatic
        fun not(expression: Expression) = expression.not()

        fun raw(expr: String): Expression = Raw(expr, emptyList())
        fun raw(expr: String, vararg args: Any): Expression = Raw(expr, bindValues(*args).toList())
        fun raw(expr: String, vararg args: String): Expression = Raw(expr, args.toList())
    }

    /**
     * The number of "dynamic" arguments in this expression. This may not be the same as the actual number of arguments
     * returned from buildSql
     */
    internal open val numOfArgs = 0

    fun args(vararg args: Any) = args(*bindValues(*args))
    open fun args(vararg args: String): Expression {
        require(args.isEmpty()) { "invalid number of arguments specified" }
        return this
    }

    infix fun and(expression: Expression): Expression = binaryExpr(Binary.AND, expression)
    infix fun or(expression: Expression): Expression = binaryExpr(Binary.OR, expression)
    open operator fun not(): Expression = Unary(Unary.NOT, this)

    infix fun eq(constant: Number): Expression = eq(constant(constant))
    infix fun eq(constant: String): Expression = eq(constant(constant))
    infix fun eq(constant: Any): Expression = eq(constant(constant))
    infix fun eq(expression: Expression): Expression = Binary(Binary.EQ, this, expression)

    infix fun lt(constant: Number): Expression = lt(constant(constant))
    infix fun lt(constant: Any): Expression = lt(constant(constant))
    infix fun lt(expression: Expression): Expression = Binary(Binary.LT, this, expression)

    infix fun lte(constant: Number): Expression = lte(constant(constant))
    infix fun lte(constant: Any): Expression = lte(constant(constant))
    infix fun lte(expression: Expression): Expression = Binary(Binary.LTE, this, expression)

    infix fun gt(constant: Number): Expression = gt(constant(constant))
    infix fun gt(constant: Any): Expression = gt(constant(constant))
    infix fun gt(expression: Expression): Expression = Binary(Binary.GT, this, expression)

    infix fun gte(constant: Number): Expression = gte(constant(constant))
    infix fun gte(constant: Any): Expression = gte(constant(constant))
    infix fun gte(expression: Expression): Expression = Binary(Binary.GTE, this, expression)

    fun oneOf(vararg expressions: Expression): Expression = Binary(Binary.IN, this, *expressions)
    fun oneOf(literals: List<Expression>): Expression = Binary(Binary.IN, this, *literals.toTypedArray())
    fun notIn(vararg expressions: Expression): Expression = Binary(Binary.NOTIN, this, *expressions)

    @Suppress("ktlint:standard:function-naming")
    fun `is`(expression: Expression): Expression = Binary(Binary.IS, this, expression)
    fun isNull(): Expression = Binary(Binary.IS, this, NULL)
    fun isNot(expression: Expression): Expression = Binary(Binary.ISNOT, this, expression)

    infix fun ne(constant: Number): Expression = ne(constant(constant))
    infix fun ne(constant: String): Expression = ne(constant(constant))
    infix fun ne(constant: Any): Expression = ne(constant(constant))
    infix fun ne(expression: Expression): Expression = Binary(Binary.NE, this, expression)

    internal open fun binaryExpr(op: String, expression: Expression): Expression = Binary(op, this, expression)

    internal abstract fun buildSql(dao: AbstractDao): QueryComponent

    @Parcelize
    data class Field internal constructor(
        private val table: Table<*>? = null,
        private val name: String,
    ) : Expression() {
        @JvmOverloads
        fun count(isDistinct: Boolean = false) = Aggregate(Aggregate.COUNT, isDistinct, this)
        @JvmOverloads
        fun max(isDistinct: Boolean = false) = Aggregate(Aggregate.MAX, isDistinct, this)
        @JvmOverloads
        fun min(isDistinct: Boolean = false) = Aggregate(Aggregate.MIN, isDistinct, this)
        @JvmOverloads
        fun sum(isDistinct: Boolean = false) = Aggregate(Aggregate.SUM, isDistinct, this)

        @Transient
        @IgnoredOnParcel
        private var sql: QueryComponent? = null

        override fun buildSql(dao: AbstractDao) =
            sql ?: QueryComponent(if (table != null) "${table.sqlPrefix(dao)}$name" else name).also { sql = it }
    }

    @Parcelize
    data class Literal internal constructor(
        private val strValue: String? = null,
        private val numValue: Number? = null,
        private val isConstant: Boolean,
    ) : Expression() {
        internal constructor(value: Number, constant: Boolean) : this(null, value, constant)
        internal constructor(value: String, constant: Boolean) : this(value, null, constant)

        override val numOfArgs get() = if (isConstant) 0 else 1

        override fun args(vararg args: String): Literal {
            require(args.size == numOfArgs) { "incorrect number of args specified" }
            return if (isConstant) this else Literal(args[0], false)
        }

        @IgnoredOnParcel
        private val sql by lazy {
            when {
                isConstant -> when {
                    numValue != null -> QueryComponent(numValue.toString())
                    strValue != null -> QueryComponent("?", strValue)
                    else -> QueryComponent("NULL")
                }

                numValue != null -> QueryComponent("?", numValue.toString())

                else -> QueryComponent("?", strValue.orEmpty())
            }
        }

        override fun buildSql(dao: AbstractDao) = sql
    }

    @Parcelize
    private data class Raw(private val expr: String, private val args: List<String>) : Expression() {
        override val numOfArgs get() = args.size

        override fun args(vararg args: String) = copy(args = args.toList())

        override fun buildSql(dao: AbstractDao) = QueryComponent(expr, *args.toTypedArray())
    }

    @Parcelize
    private data class Binary(private val op: String, private val exprs: List<Expression>) : Expression() {
        companion object {
            internal const val LT = "<"
            internal const val LTE = "<="
            internal const val GT = ">"
            internal const val GTE = ">="
            internal const val EQ = "=="
            internal const val NE = "!="
            internal const val IS = "IS"
            internal const val ISNOT = "IS NOT"
            internal const val IN = "IN"
            internal const val NOTIN = "NOT IN"
            internal const val AND = "AND"
            internal const val OR = "OR"
        }

        constructor(op: String, vararg exprs: Expression) : this(op, exprs.toList())

        @Transient
        @IgnoredOnParcel
        override val numOfArgs = exprs.sumOf { it.numOfArgs }

        override fun args(vararg args: String): Expression {
            require(args.size == numOfArgs) { "incorrect number of args specified" }
            if (args.isEmpty()) return this

            var pos = 0
            val exprs = exprs.map {
                when (it.numOfArgs) {
                    0 -> it
                    else -> it.args(*args.sliceArray(pos until pos + it.numOfArgs)).also { pos += it.numOfArgs }
                }
            }

            return Binary(op, *exprs.toTypedArray())
        }

        override fun not() = when (op) {
            EQ -> Binary(NE, *exprs.toTypedArray())
            NE -> Binary(EQ, *exprs.toTypedArray())
            IS -> Binary(ISNOT, *exprs.toTypedArray())
            ISNOT -> Binary(IS, *exprs.toTypedArray())
            IN -> Binary(NOTIN, *exprs.toTypedArray())
            NOTIN -> Binary(IN, *exprs.toTypedArray())
            else -> super.not()
        }

        override fun binaryExpr(op: String, expression: Expression) = when {
            this.op != op -> super.binaryExpr(op, expression)

            // chain binary expressions together when possible
            op == AND || op == OR -> Binary(op, *exprs.toTypedArray(), expression)

            else -> super.binaryExpr(op, expression)
        }

        @Transient
        @IgnoredOnParcel
        private var sql: QueryComponent? = null
        private fun generateSql(dao: AbstractDao): QueryComponent {
            val query = QueryComponent("(") + when (op) {
                IN, NOTIN -> {
                    exprs[0].buildSql(dao) + " $op (" + exprs.subList(1, exprs.size)
                        .joinToQueryComponent(",") { it.buildSql(dao) } + ")"
                }

                else -> exprs.joinToQueryComponent(" $op ") { it.buildSql(dao) }
            } + ")"
            sql = query
            return query
        }

        override fun buildSql(dao: AbstractDao) = sql ?: generateSql(dao)
    }

    @Parcelize
    private data class Unary(private val op: String, private val expr: Expression) : Expression() {
        companion object {
            internal const val NOT = "NOT"
        }

        override val numOfArgs get() = expr.numOfArgs

        override fun args(vararg args: String) = copy(expr = expr.args(*args))
        override fun not() = when (op) {
            NOT -> expr
            else -> super.not()
        }

        @Transient
        @IgnoredOnParcel
        private var sql: QueryComponent? = null
        private fun generateSql(dao: AbstractDao) =
            (QueryComponent("$op (") + expr.buildSql(dao) + ")").also { sql = it }

        override fun buildSql(dao: AbstractDao) = sql ?: generateSql(dao)
    }

    @Parcelize
    data class Aggregate internal constructor(
        private val op: String,
        private val isDistinct: Boolean,
        private val expr: Field,
    ) : Expression() {
        companion object {
            internal const val COUNT = "COUNT"
            internal const val MAX = "MAX"
            internal const val MIN = "MIN"
            internal const val SUM = "SUM"
        }

        override val numOfArgs get() = expr.numOfArgs

        override fun args(vararg args: String) = copy(expr = expr.args(*args) as Field)

        fun distinct(isDistinct: Boolean) = copy(isDistinct = isDistinct)

        @Transient
        @IgnoredOnParcel
        private var sql: QueryComponent? = null

        // {mOp} (DISTINCT {expr})
        private fun generateSql(dao: AbstractDao) =
            (QueryComponent("$op (${if (isDistinct) "DISTINCT " else ""}") + expr.buildSql(dao) + ")")
                .also { sql = it }

        override fun buildSql(dao: AbstractDao) = sql ?: generateSql(dao)
    }
}
